### PR TITLE
Add batchify transformer to help end to end models

### PR DESCRIPTION
**Pull request by sandeep-krishnamurthy**
## Description ##
Add new transformer in Gluon "Batchify". This transformer joins/batchify a list of tensors of shape (C x H x W) into a single tensor of shape (N x C x H x W).

This will be helpful when users have created a fused end to end model (transformers + network), saved the graph and this end to end model is deployed in production for "single request inference" or "mini-batch inference". The raw input can be of different shapes, till the input reaches resize transformation, if running a mini batch inference, then input will be a list of NDArrays. After Resize, these individual NDArrays will batched with this transformer.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Add new transformer Batchify to take list of NDArrays and batchify them in to 1 NDArray

@stu1130

### CODE REVIEW

**Comment by sandeep-krishnamurthy** [comment]
> This seems to be the concern of batchify function in dataloader.

Added the description explaining the usage. Basically this will be helpful during mini batch inference on an end to end model i.e., a fused model (transformer + network).

Input can be different shapes. Resize transformer is also part of the network.
Example model looks like - Resize -> Batchify -> ToTensor -> Normalize -> Network
Example mini batch inference looks like
```
im1 = imread(...)
im2 = imread(...)
Outputs = executor.forward(data0=im1, data1=im2)
```

**Comment by stu1130** [comment]
@mxnet-label-bot add [pr-awaiting-review]
Great work! @sandeep-krishnamurthy

File: python/mxnet/gluon/data/vision/transforms.py [hunk 1]
```diff
@@ -485,3 +485,52 @@ def __init__(self, alpha):
...
+    This transformer is useful when transformation pipeline is fused into
+    neural network graph resulting in single model/graph. When running
+    mini batch inference on such graph, raw input tensors, before transformations,
+    can be of different shapes, hence, cannot be batched into single tensor. Hence,
+    input to the graph will be list of Tensors that can be batched after Resize
+    transformation.
+
+    For example, a typical graph can look like below:
+
+    Imdecode -> Resize -> Batchify -> ToTensor -> Normalize -> Network
```
Comment (zhreshold) [inline python/mxnet/gluon/data/vision/transforms.py]: This argument does not sound right, images after resize has layout HWC, which violate the input expectation of CHW with this Batchify block

Comment (szha) [reply python/mxnet/gluon/data/vision/transforms.py]: More importantly, dataset.transform takes care of only sample-level transformation. It needs to be clear which class is responsible for stacking the samples. This is currently done in dataloader.

Comment (sandeep-krishnamurthy) [reply python/mxnet/gluon/data/vision/transforms.py]: @zhreshold - Thanks for pointing it out. Functionally, input can be either HWC or CHW it gets converted to NHWC or NCHW respectively. Will update the code comment.

Comment (sandeep-krishnamurthy) [reply python/mxnet/gluon/data/vision/transforms.py]: @szha - The way I am thinking is - this transformer is used when user is exporting a model for inference and he wants to fuse transformations as part of the network. This transformer is not applicable for training / validation which all works as existing today.
So to summarize, the idea here is:
1. Training does not changing anything from what it is today - dataset and dataloader
2. When user wants to export a model transformations (Resize, Batchify, ToTensor, Normalize) is concatenated to network block and exported.

Comment (sandeep-krishnamurthy) [reply python/mxnet/gluon/data/vision/transforms.py]: I agree, but, thinking what is more intuitive and ease of use.

Comment (szha) [reply python/mxnet/gluon/data/vision/transforms.py]: Introducing this would only make it confusing to people who are already using the dataloader in python. If the goal is to make it available to other users, one simple approach would be to wrap this stack operator in the logic of exporting data preprocessing pipeline. @zhreshold

File: tests/python/unittest/test_gluon_data_vision.py [hunk 2]
```diff
@@ -83,6 +83,25 @@ def test_transformer():
     transform(mx.nd.ones((245, 480, 3), dtype='uint8')).wait_to_read()
 
 
+@with_seed()
+def test_batchify():
+    transformer = transforms.Batchify()
+    im1 = nd.random.uniform(0, 1, (3, 28, 28))
+    im2 = nd.random.uniform(0, 1, (3, 28, 28))
+    im3 = nd.random.uniform(0, 1, (3, 28, 28))
```
Comment (vandanavk) [inline tests/python/unittest/test_gluon_data_vision.py]: Could you add a test with inputs of different sizes too? That may serve as an example for users as well

**Comment by ankkhedia** [comment]
@sandeep-krishnamurthy Could you please look into the review comments above?

File: tests/python/unittest/test_gluon_data_vision.py (hunk 2 above)
Comment (anirudhacharya) [reply tests/python/unittest/test_gluon_data_vision.py]: can the batch sizes be randomized?

**Comment by sandeep-krishnamurthy** [comment]
Closing this PR because having is transformer is more confusing than useful for users. As suggested above in reviews, such transformer should be internally managed by respective export/import APIs.
